crypto/x509/pcy_node.c

```diff
@@ -93,27 +93,28 @@ X509_POLICY_NODE *ossl_policy_level_add_node(X509_POLICY_LEVEL *level,
         }
     }
 
+    tree->node_count++;
+    if (parent)
+        parent->nchild++;
+
     if (extra_data) {
         if (tree->extra_data == NULL)
             tree->extra_data = sk_X509_POLICY_DATA_new_null();
         if (tree->extra_data == NULL) {
             ERR_raise(ERR_LIB_X509V3, ERR_R_CRYPTO_LIB);
-            goto node_error;
+            goto extra_data_error;
```

Comment:
this is not right if level == NULL, i.e. when this is called from here:
https://github.com/openssl/openssl/blob/6dea91f56dcbcb0979dd36790664808ad960faf9/crypto/x509/pcy_tree.c#L568-L569
but OTOH, when this error happens, this allocation is not free'd:
https://github.com/openssl/openssl/blob/6dea91f56dcbcb0979dd36790664808ad960faf9/crypto/x509/pcy_tree.c#LL562C13-L562C13

Reply:
actually I'd rather undo all changes when this error happens here?

Reply:
Good catch, I'll push a change that will free `node` if it wasn't added to a level.

I believe the missing free you're pointing to should be the responsibility of the caller. Other call sites of `ossl_policy_level_add_node()` always call `ossl_policy_data_free(data)` on error, see for example https://github.com/openssl/openssl/blob/master/crypto/x509/pcy_tree.c#L195

Reply:
I added a commit on top with error handling for this invocation.

Reply:
Thanks, but could you instead pop the node off the stack when the subsequent allocation fails,
so that this function guarantees not to change anything when it is failing, instead if doing just
half of it's job?

Reply:
Done.